packages/@n8n/task-runner-node-js/src/task-runner.ts

```diff
@@ -0,0 +1,364 @@
+	) {
+		this.id = nanoid();
+
+		const url = new URL(wsUrl);
+		url.searchParams.append('id', this.id);
+		this.ws = new WebSocket(url.toString(), {
+			headers: {
+				authorization: `Bearer ${grantToken}`,
+			},
+		});
+		console.log('1');
+		this.ws.addEventListener('message', this._wsMessage);
+		this.ws.addEventListener('close', this.stopTaskOffers);
+	}
+
+	private _wsMessage = (message: MessageEvent) => {
+		// eslint-disable-next-line n8n-local-rules/no-uncaught-json-parse
+		const data = JSON.parse(message.data as string) as N8nMessage.ToRunner.All;
+		void this.onMessage(data);
+	};
+
+	private stopTaskOffers = () => {
+		this.canSendOffers = false;
+		if (this.offerInterval) {
+			clearInterval(this.offerInterval);
+			this.offerInterval = undefined;
+		}
+	};
+
+	private startTaskOffers() {
+		this.canSendOffers = true;
+		if (this.offerInterval) {
+			clearInterval(this.offerInterval);
+		}
+		this.offerInterval = setInterval(() => this.sendOffers(), 250);
+	}
+
+	deleteStaleOffers() {
+		for (const key of Object.keys(this.openOffers)) {
+			if (this.openOffers[key].validUntil < process.hrtime.bigint()) {
+				delete this.openOffers[key];
+			}
+		}
+	}
+
+	sendOffers() {
+		this.deleteStaleOffers();
+
+		const offersToSend =
+			this.maxConcurrency -
+			(Object.values(this.openOffers).length + Object.values(this.runningTasks).length);
+
+		if (offersToSend > 0) {
+			for (let i = 0; i < offersToSend; i++) {
+				const offer: TaskOffer = {
+					offerId: nanoid(),
+					validUntil:
+						process.hrtime.bigint() + BigInt((VALID_TIME_MS + VALID_EXTRA_MS) * 1_000_000), // Adding a little extra time to account for latency
+				};
+				this.openOffers[offer.offerId] = offer;
+				this.send({
+					type: 'runner:taskoffer',
+					taskType: this.taskType,
+					offerId: offer.offerId,
+					validFor: VALID_TIME_MS,
+				});
+			}
+		}
+	}
+
+	send(message: RunnerMessage.ToN8n.All) {
+		this.ws.send(JSON.stringify(message));
+	}
+
+	onMessage(message: N8nMessage.ToRunner.All) {
+		console.log({ message });
+		switch (message.type) {
+			case 'broker:inforequest':
+				this.send({
+					type: 'runner:info',
+					name: this.name ?? 'Node.js Task Runner SDK',
+					types: [this.taskType],
+				});
+				break;
+			case 'broker:runnerregistered':
+				this.startTaskOffers();
+				break;
+			case 'broker:taskofferaccept':
+				this.offerAccepted(message.offerId, message.taskId);
+				break;
+			case 'broker:taskcancel':
+				this.taskCancelled(message.taskId);
+				break;
+			case 'broker:tasksettings':
+				void this.receivedSettings(message.taskId, message.settings);
+				break;
+			case 'broker:taskdataresponse':
+				this.processDataResponse(message.requestId, message.data);
+				break;
+			case 'broker:rpcresponse':
+				this.handleRpcResponse(message.callId, message.status, message.data);
+		}
+	}
+
+	processDataResponse(requestId: string, data: unknown) {
+		const request = this.dataRequests[requestId];
+		if (!request) {
```

Comment:
Basically never, it's basically just there for typechecking

Reply:
In other places (like the broker and manager), they should basically never happen. It's more just in case something else goes wrong (busy event loop), and it's getting messages that are already resolved elsewhere. I'd considered logging them, but I thought it'd either never get hit, or be really noisy in some environments.